File: typescript/src/challenges/teach-back.ts
/**
 * Teach-back challenge for CRITICAL-risk actions.
 *
 * The operator must explain, in their own words, what the action will do
 * and what its effects are. This is the strongest single-person challenge
 * because it requires active comprehension rather than passive recognition.
 */

import * as pathModule from "node:path";

import {
  type ActionContext,
  type ChallengeProtocol,
  type ChallengeResult,
  ChallengeType,
  type RiskAssessment,
  createChallengeResult,
  describeAction,
} from "../types.js";

import {
  type TeachBackValidator,
  KeywordValidator,
} from "./validators.js";

// ---------------------------------------------------------------------------
// Key term extraction
// ---------------------------------------------------------------------------

/**
 * Derive a set of key terms from the action context.
 *
 * Key terms are the building blocks used to check whether the operator's
 * free-text explanation demonstrates genuine comprehension. They come from:
 *
 * - Parts of the function name (split on `_`, `-`, and camelCase boundaries).
 * - Significant keyword-argument values (strings longer than 2 chars,
 *   path basenames, table names, etc.).
 * - Significant positional-argument values using the same heuristics.
 */
function extractKeyTerms(ctx: ActionContext): string[] {
  const terms: string[] = [];

  // Function name parts
  const nameParts = ctx.functionName
    .replace(/([a-z])([A-Z])/g, "$1 $2")
    .replace(/-/g, " ")
    .replace(/_/g, " ");
  for (const word of nameParts.split(/\s+/)) {
    const cleaned = word.trim().toLowerCase();
    if (cleaned.length > 2) {
      terms.push(cleaned);
    }
  }

  // Significant argument values
  const allValues: unknown[] = [
    ...ctx.args,
    ...Object.values(ctx.kwargs),
  ];
  for (const val of allValues) {
    if (typeof val === "string" && val.length > 2) {
      // Try to extract the basename of a path
      if (val.includes("/") || val.includes("\\")) {
        terms.push(pathModule.basename(val).toLowerCase());
      }
      // Short-ish strings are likely identifiers or names
      if (val.length <= 80) {
        terms.push(val.trim().toLowerCase());
      }
    } else if (typeof val === "number") {
      terms.push(String(val));
    }
  }

  // Deduplicate while preserving order
  const seen = new Set<string>();
  const unique: string[] = [];
  for (const t of terms) {
    if (!seen.has(t)) {
      seen.add(t);
      unique.push(t);
    }
  }
  return unique;
}

// ---------------------------------------------------------------------------
// TeachBackChallenge options
// ---------------------------------------------------------------------------

/**
 * Options for configuring the TeachBackChallenge.
 */
export interface TeachBackChallengeOptions {
  /** Minimum word count for the operator's explanation. Default 15. */
  minWords?: number;

  /**
   * Wall-clock seconds the action summary must be visible before the
   * prompt becomes active. Default 30.
   */
  minReviewSeconds?: number;

  /**
   * Pluggable validator for teach-back explanations. When provided, replaces
   * the default keyword-based validation. See `KeywordValidator` or implement
   * the `TeachBackValidator` interface.
   */
  validator?: TeachBackValidator;

  /** Custom input function. */
  promptFn?: (prompt: string) => Promise<string>;

  /** Custom output function. */
  printFn?: (message: string) => void;
}

// ---------------------------------------------------------------------------
// TeachBackChallenge
// ---------------------------------------------------------------------------

/**
 * Teach-back verification for CRITICAL-risk actions.
 *
 * The operator is shown the full action details and asked to explain, in
 * their own words, what the action will do and what its effects are.
 *
 * Validation rules:
 *
 * 1. The response must be at least `minWords` words long.
 * 2. The response must contain at least one key term extracted from
 *    the action context (function name parts, significant arg values).
 * 3. (Optional) A custom validator can apply additional checks.
 */
export class TeachBackChallenge implements ChallengeProtocol {
  readonly minWords: number;
  readonly minReviewSeconds: number;
  private readonly _validator: TeachBackValidator;
  private readonly _promptFn: (prompt: string) => Promise<string>;
  private readonly _printFn: (message: string) => void;

  constructor(options: TeachBackChallengeOptions = {}) {
    this.minWords = options.minWords ?? 15;
    this.minReviewSeconds = options.minReviewSeconds ?? 30.0;
    this._promptFn = options.promptFn ?? defaultPrompt;
    this._printFn = options.printFn ?? defaultPrint;

    this._validator = options.validator ?? new KeywordValidator({ minWords: this.minWords });
  }

  get challengeType(): ChallengeType {
    return ChallengeType.TEACH_BACK;
  }

  // -- presentation -------------------------------------------------------

  /**
   * Present the teach-back challenge to the operator.
   */
  async present(
    ctx: ActionContext,
    risk: RiskAssessment
  ): Promise<ChallengeResult> {
    const start = performance.now();
    const print = this._printFn;
    const prompt = this._promptFn;

    // -- render full action details -------------------------------------
    const separator = "=".repeat(60);
    print(`\n${separator}`);
    print("  TEACH-BACK CHALLENGE  --  CRITICAL RISK ACTION");
    print(separator);
    print(`  Action:      ${ctx.functionName}`);
    print(
      `  Risk:        ${risk.level.toUpperCase()} (${risk.score.toFixed(2)})`
    );
    if (ctx.functionDoc) {
      print(`  Description: ${ctx.functionDoc}`);
    }
    print(`  Call:        ${describeAction(ctx)}`);
    if (ctx.args.length > 0) {
      print(`  Positional:  ${JSON.stringify(ctx.args)}`);
    }
    if (Object.keys(ctx.kwargs).length > 0) {
      print(`  Keyword:     ${JSON.stringify(ctx.kwargs)}`);
    }
    if (risk.factors.length > 0) {
      print("  Risk factors:");
      for (const factor of risk.factors) {
        print(`    - ${factor.name}: ${factor.description}`);
      }
    }
    print(separator);

    // -- enforce minimum review time ------------------------------------
    const elapsed = (performance.now() - start) / 1000;
    if (elapsed < this.minReviewSeconds) {
      const remaining = this.minReviewSeconds - elapsed;
      print(
        `  [Read carefully. Prompt activates in ${Math.round(remaining)}s...]`
      );
      await sleep(remaining);
    }

    // -- collect free-text explanation -----------------------------------
    print(
      "\n  In your own words, explain what this action will do and " +
        "what its effects are:"
    );
    const explanation = (await prompt("  > ")).trim();

    // -- validate via pluggable validator ---------------------------------
    const validatorResult = await this._validator.validate(explanation, ctx);
    const passed = validatorResult.passed;
    const validationNotes: string[] = [validatorResult.notes];

    // Extract details for backward-compatible result structure
    const keyTerms = extractKeyTerms(ctx);
    const explanationLower = explanation.toLowerCase();
    const matchedTerms = keyTerms.filter((t) =>
      explanationLower.includes(t)
    );
    const wordCount = explanation.split(/\s+/).filter(Boolean).length;
    const totalElapsed = (performance.now() - start) / 1000;

    const status = passed ? "PASSED" : "FAILED";
    print(`\n  Teach-back result: ${status}`);
    for (const note of validationNotes) {
      print(`    - ${note}`);
    }

    return createChallengeResult({
      passed,
      challengeType: this.challengeType,
      responseTimeSeconds: totalElapsed,
      questionsAsked: 1,
      questionsCorrect: passed ? 1 : 0,
      details: {
        explanation,
        wordCount,
        keyTerms,
        matchedTerms,
        passed,
        validationNotes,
      },
    });
  }
}

// ---------------------------------------------------------------------------
// Default I/O helpers
// ---------------------------------------------------------------------------

function defaultPrint(message: string): void {
  // eslint-disable-next-line no-console
  console.log(message);
}

async function defaultPrompt(message: string): Promise<string> {
  const { createInterface } = await import("node:readline");
  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise<string>((resolve) => {
    rl.question(message, (answer) => {
      rl.close();
      resolve(answer);
    });
  });
}

function sleep(seconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

File: python/src/attesta/challenges/teach_back.py
"""Teach-back challenge for CRITICAL-risk actions.

The operator must explain, in their own words, what the action will do
and what its effects are.  This is the strongest single-person challenge
because it requires *active* comprehension rather than passive recognition.
"""

from __future__ import annotations

import asyncio
import time
from typing import Any

from attesta.core.types import (
    ActionContext,
    ChallengeResult,
    ChallengeType,
    RiskAssessment,
)
from attesta.challenges.validators import _extract_key_terms


class TeachBackChallenge:
    """Teach-back verification for CRITICAL-risk actions.

    The operator is shown the full action details and asked to explain, in
    their own words, what the action will do and what its effects are.

    Validation rules:

    1. The response must be at least *min_words* words long.
    2. The response must contain at least one **key term** extracted from
       the action context (function name parts, significant arg values).
    3. *(Optional)* If a custom ``validator`` is provided, it can apply
       additional comprehension checks.

    Parameters
    ----------
    min_words:
        Minimum word count for the operator's explanation.
    min_review_seconds:
        Wall-clock seconds the action summary must be visible before the
        prompt becomes active.
    validator:
        Pluggable validator implementing ``TeachBackValidator`` protocol.
        Defaults to ``KeywordValidator(min_words=min_words)``.
    """

    def __init__(
        self,
        min_words: int = 15,
        min_review_seconds: float = 30.0,
        validator: Any = None,
    ) -> None:
        self.min_words = min_words
        self.min_review_seconds = min_review_seconds

        if validator is not None:
            self._validator = validator
        else:
            from attesta.challenges.validators import KeywordValidator
            self._validator = KeywordValidator(min_words=min_words)

    @property
    def challenge_type(self) -> ChallengeType:
        return ChallengeType.TEACH_BACK

    # -- presentation -----------------------------------------------------

    async def present(
        self, ctx: ActionContext, risk: RiskAssessment
    ) -> ChallengeResult:
        """Present the teach-back challenge to the operator."""
        start = time.monotonic()
        loop = asyncio.get_running_loop()

        # -- render full action details -----------------------------------
        separator = "=" * 60
        print(f"\n{separator}")
        print(f"  TEACH-BACK CHALLENGE  --  CRITICAL RISK ACTION")
        print(f"{separator}")
        print(f"  Action:      {ctx.function_name}")
        print(f"  Risk:        {risk.level.value.upper()} ({risk.score:.2f})")
        if ctx.function_doc:
            print(f"  Description: {ctx.function_doc}")
        print(f"  Call:        {ctx.description}")
        if ctx.args:
            print(f"  Positional:  {ctx.args!r}")
        if ctx.kwargs:
            print(f"  Keyword:     {ctx.kwargs!r}")
        if risk.factors:
            print(f"  Risk factors:")
            for factor in risk.factors:
                print(f"    - {factor.name}: {factor.description}")
        print(f"{separator}")

        # -- enforce minimum review time ----------------------------------
        elapsed = time.monotonic() - start
        if elapsed < self.min_review_seconds:
            remaining = self.min_review_seconds - elapsed
            print(f"  [Read carefully. Prompt activates in {remaining:.0f}s...]")
            await asyncio.sleep(remaining)

        # -- collect free-text explanation --------------------------------
        print(
            "\n  In your own words, explain what this action will do and "
            "what its effects are:"
        )

        def _read_input() -> str:
            try:
                return input("  > ").strip()
            except (EOFError, KeyboardInterrupt):
                return ""

        explanation: str = await loop.run_in_executor(None, _read_input)

        # -- validate via pluggable validator -----------------------------
        passed, validation_note = await self._validator.validate(explanation, ctx)
        validation_notes: list[str] = [validation_note]

        # Extract details for backward-compatible result structure
        key_terms = _extract_key_terms(ctx)
        explanation_lower = explanation.lower()
        matched_terms = [t for t in key_terms if t in explanation_lower]
        word_count = len(explanation.split())
        elapsed = time.monotonic() - start

        status = "PASSED" if passed else "FAILED"
        print(f"\n  Teach-back result: {status}")
        for note in validation_notes:
            print(f"    - {note}")

        return ChallengeResult(
            passed=passed,
            challenge_type=self.challenge_type,
            response_time_seconds=elapsed,
            questions_asked=1,
            questions_correct=1 if passed else 0,
            details={
                "explanation": explanation,
                "word_count": word_count,
                "key_terms": key_terms,
                "matched_terms": matched_terms,
                "passed": passed,
                "validation_notes": validation_notes,
            },
        )
